component/module/git/git.go

```diff
@@ -0,0 +1,268 @@
+	*args = DefaultArguments
+
+	type arguments Arguments
+	return f((*arguments)(args))
+}
+
+// Component implements the module.git component.
+type Component struct {
+	opts component.Options
+	log  log.Logger
+	mod  *module.ModuleComponent
+
+	mut      sync.RWMutex
+	repo     *vcs.GitRepo
+	repoOpts vcs.GitRepoOptions
+	args     Arguments
+
+	argsChanged chan struct{}
+
+	healthMut sync.RWMutex
+	health    component.Health
+}
+
+var (
+	_ component.Component       = (*Component)(nil)
+	_ component.HealthComponent = (*Component)(nil)
+	_ component.HTTPComponent   = (*Component)(nil)
+)
+
+// New creates a new module.git component.
+func New(o component.Options, args Arguments) (*Component, error) {
+	c := &Component{
+		opts: o,
+		log:  o.Logger,
+
+		mod: module.NewModuleComponent(o),
+
+		argsChanged: make(chan struct{}, 1),
+	}
+
+	if err := c.Update(args); err != nil {
+		return nil, err
+	}
+	return c, nil
+}
+
+// Run implements component.Component.
+func (c *Component) Run(ctx context.Context) error {
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+
+	go c.mod.RunFlowController(ctx)
+
+	var (
+		ticker  *time.Ticker
+		tickerC <-chan time.Time
+	)
+
+	for {
+		select {
+		case <-ctx.Done():
+			return nil
+
+		case <-c.argsChanged:
+			c.mut.Lock()
+			{
+				level.Info(c.log).Log("msg", "updating repository pull frequency", "new_frequency", c.args.PullFrequency)
+
+				if c.args.PullFrequency > 0 {
+					if ticker == nil {
+						ticker = time.NewTicker(c.args.PullFrequency)
+						tickerC = ticker.C
+					} else {
+						ticker.Reset(c.args.PullFrequency)
+					}
+				} else {
+					if ticker != nil {
+						ticker.Stop()
+					}
+					ticker = nil
+					tickerC = nil
+				}
+			}
+			c.mut.Unlock()
+
+		case <-tickerC:
+			level.Info(c.log).Log("msg", "updating repository", "new_frequency", c.args.PullFrequency)
+			c.tickPollFile(ctx)
+		}
+	}
+}
+
+func (c *Component) tickPollFile(ctx context.Context) {
+	c.mut.Lock()
+	err := c.pollFile(ctx, c.args)
+	c.mut.Unlock()
+
+	c.updateHealth(err)
+}
+
+func (c *Component) updateHealth(err error) {
+	c.healthMut.Lock()
+	defer c.healthMut.Unlock()
+
+	if err != nil {
+		c.health = component.Health{
+			Health:     component.HealthTypeUnhealthy,
+			Message:    err.Error(),
+			UpdateTime: time.Now(),
+		}
+	} else {
+		c.health = component.Health{
+			Health:     component.HealthTypeHealthy,
+			Message:    "module updated",
+			UpdateTime: time.Now(),
+		}
+	}
+}
+
+// Update implements component.Component.
+func (c *Component) Update(args component.Arguments) (err error) {
+	defer func() {
+		c.updateHealth(err)
+	}()
+
+	c.mut.Lock()
+	defer c.mut.Unlock()
+
+	newArgs := args.(Arguments)
+
+	// TODO(rfratto): store in a repo-specific directory so changing repositories
+	// doesn't risk break the module loader if there's a SHA collision between
+	// the two different repositories.
+	repoPath := filepath.Join(c.opts.DataPath, "repo")
```

Comment:
I could try to fix this before merging, but honestly I'm not sure if it's necessary. I'm tempted to leave this as-is and see if it ever becomes a problem for anyone.